libevmasm/AssemblyItem.cpp

```diff
@@ -235,6 +245,11 @@ size_t AssemblyItem::returnValues() const
 		return std::get<1>(*m_verbatimBytecode);
 	case AuxDataLoadN:
 		return 1;
+	case CallF:
+		return functionSignature().retsNum;
+	case JumpF:
+	case RetF:
+		return 0;
```

Comment:
I think it's `RETF` that returns, not `CALLF` :)

```suggestion
	case RetF:
		return static_cast<size_t>(data());
	case JumpF:
	case CallF:
		return 0;
```

Reply:
BTW, we should add an assert against `m_data` being null in `data()`. Currently it's not checked.

Reply:
Added. Good point. Why do the tests pass then?

Reply:
Wait. But if we make `CallF` to return 0 we would have stack height mismatch between assembly stack and `OptimizedEVMCodeTransform::m_stack`. IMO it has to return number of return for this function signature in `returnValues`.

Reply:
As `JumpF` also can be used to jump to returning function the same applies to it.

Reply:
And `RetF` should only get arguments according to number of returns from the function and nothing return to th stack.

Reply:
Oh, I see. I was thinking about arguments and return values here as items pushed/popped from the stack by the instruction, but apparently this is describing the effects of the whole call.

So for `JumpF` and `CallF` we want `argsNum` for `arguments()` and `retsNum` for `returnValues()` (or `0` if non-returning). It should basically match the inputs/outputs of code section.

Reply:
Not sure about `RetF` though. Shouldn't it be defined as not taking and not returning anything? I.e. be grouped together with `Operation` and match the number of slots popped/pushed by its instruction?

Inside the code section we're supposed to have instructions that consume `I` inputs and produce `O` outputs, giving you `O - I` as delta. If you define `RETF` as non-neutral, consuming `O` items, the delta will be just `-I`, which seems wrong.

Reply:
Yes. I though that `RetF` should consume its arguments from the stack but they have to be left there for the function caller.

Reply:
ok, it's correct now, but you should also move `RetF` to the branch that handles `hasInstruction()`/`Operation` so that the zeros come automatically from instruction definition.